### PR TITLE
D8CORE-1349 Added visual indicator for links to unpublished pages

### PR DESCRIPTION
**Pull request by pookmish**
This needs some front end attention.

### CODE REVIEW

**Comment by pookmish** [comment]
I'm going to pass this off to @jenbreese 
Please let me know if you need anything from me on this.

**Comment by jenbreese** [comment]
# READY FOR REVIEW

# Summary
- This styles the indicators pointing out the unpublished files.

# Review By (Date)
- Soonish

# Urgency
- Low

# Steps to Test

1. Pull in this branch.
2. npm run publish
3. Clear caches.
4. Log in.
5. Unpublish a page.
6. Verify you get the the indicator.
7. Try on on main and secondary navigation pages.

# Affected Projects or Products
- D8CORE-1349

# Associated Issues and/or People
- D8CORE-1349

- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- @pookmish 

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

**Comment by pookmish** [comment]
Cool. thanks @jenbreese this looks good to me. @sherakama care to take a look?

File: src/scss/admin/_menu.scss [hunk 1]
```diff
@@ -0,0 +1,37 @@
+@charset "UTF-8";
+
+nav {
+  .unpublished-indicator {
+    @include padding(4px 10px);
+
+    content: "Unpublished Page";
```
Comment (sherakama) [inline src/scss/admin/_menu.scss]: It is not recommended to add text in this manner as it doesn't get translated. In this case, it doesn't matter due to the authenticated user requirement, but for future reference, added text through CSS has a negative impact on SEO and accessibility.

File: src/scss/admin/_menu.scss [hunk 2]
```diff
@@ -0,0 +1,37 @@
+@charset "UTF-8";
+
+nav {
```
Comment (sherakama) [inline src/scss/admin/_menu.scss]: I don't think you need to next the unpublished-indicator in a nav selector. You can get away with just the css class selector.

File: templates/menus/macros/nav-menu.twig [hunk 3]
```diff
@@ -28,7 +28,12 @@
     {% endif %}
 
     <li{{ list_attributes }}>
-      <a{{ link_attributes }}>{{ item.title }}</a>
+      <a{{ link_attributes }}>
+        {{ item.title }}
+        {% if item.unpublished %}
+          <div class="unpublished-indicator">Unpublished Page</div>
```
Comment (sherakama) [inline templates/menus/macros/nav-menu.twig]: Strings should be run through the `trans` filter.

File: templates/menus/macros/secondary-nav-menu.twig [hunk 4]
```diff
@@ -30,7 +30,12 @@
     {# Skip non-trail items below the first level. #}
     {% if item.in_active_trail == true or menu_level == 1 or parent.in_active_trail == true %}
       <li{{ list_attributes }}>
-        <a{{ link_attributes }}>{{ item.title }}</a>
+        <a{{ link_attributes }}>
+          {{ item.title }}
+          {% if item.unpublished %}
+            <div class="unpublished-indicator">Unpublished Page</div>
```
Comment (sherakama) [inline templates/menus/macros/secondary-nav-menu.twig]:
```suggestion
            <div class="unpublished-indicator">{{ "Unpublished Page"|t }}</div>
```

Comment (jenbreese) [reply templates/menus/macros/secondary-nav-menu.twig]: Ok. I can do that.

File: src/scss/admin/_menu.scss [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      border-left: 10px solid transparent;
+      border-right: 10px solid transparent;
+      border-top: 10px solid $su-color-blue;
+      position: absolute;
+      bottom: -10px;
+      right: 90px;
+    }
+  }
+
+  &.su-secondary-nav {
```
Comment (sherakama) [inline src/scss/admin/_menu.scss]: No need to nest this in a nav component and have a strong `nav.su-secondary-nav .unpublished-indicator` selector output. `.su-secondary-nav .unpublished-indicator` should suffice.

File: src/scss/admin/_menu.scss (hunk 1 above)
Comment (jenbreese) [reply src/scss/admin/_menu.scss]: Ok. I'll work on that tomorrow.

File: stanford_basic.theme [hunk 6]
```diff
@@ -48,7 +48,7 @@ function stanford_basic_preprocess_html(&$variables) {
     if (!empty($alias)) {
       $name = str_replace('/', '-', $alias);
       $variables['attributes']['class'][] = 'page-' . $name;
-      list($section,) = explode('/', $alias, 2);
+      [$section,] = explode('/', $alias, 2);
```
Comment (sherakama) [inline stanford_basic.theme]: ooof. I can see where this short syntax can get a little ugly. That hanging comma bugs me.

File: src/scss/admin/_menu.scss (hunk 5 above)
Comment (jenbreese) [reply src/scss/admin/_menu.scss]: Ok. I'll take a look.

**Comment by sherakama** [comment]
I really like the idea of this work as having a visual indicator to show which items are unpublished is super helpful. I left a few code comments above but also have a couple comments about the visual display. The placement of the tooltip seems obtrusive and jarring. The overlay and positioning appear way out of the natural flow of content and really catch my eye. Perhaps @kerri-augenstein has some feedback on the visual appearance of the notice. 

Screen Shot 2020-03-17 at 10 08 14 PM

Screen Shot 2020-03-17 at 10 08 30 PM

**Comment by kerri-augenstein** [comment]
@sherakama, thanks for commenting on this!! I think what we're going for is for these pages to be really, really obvious in the display, like an alert. So I'm not too bothered that the tooltip is obtrusive, and as intended catches your eye, yey, goal! But. 

Maybe we could have this tooltip only appear for a duration of seconds, 3, or 5??? And then reappear on-hover? The link will be paired with a faint blue background behind the themselves as well, so that'll help extend the visual indication, after the tooltip has faded (though the light blue is light, we may need to test how visible this ends up being for our users).

**Comment by jenbreese** [comment]
Ok, I fixed up most of the comments from @sherakama and @kerri-augenstein. I used an animation for the indicator to show up. I'm working on the hover but it isn't right yet. So thoughts on that appreciated.

**Comment by jenbreese** [comment]
@sherakama or @pookmish want to take another look?

**Comment by pookmish** [comment]
I like that animation and the fade away after some time. it helps to identify but isn't obtrusive to the experience.

**Comment by sherakama** [comment]
This looks good. Just one visual oddity.

**Comment by pookmish** [comment]
blame @sherakama https://github.com/SU-SWS/stanford_basic/blob/e911bc914300752f30488a5798f918e9ebb50cd7/templates/node.html.twig#L37

**Comment by jenbreese** [comment]
@sherakama I thought that was intentional. Would you or @pookmish like me to style it?

**Comment by pookmish** [comment]
@jenbreese i think it was intentional when Shea initially created the theme back in 2016 :laughing: 
```
$ git log -S node--unpublished
commit 2f3310139a077a60b00281efcfeef95c2fa6d8d7
Author: Shea McKinney <sherakama@gmail.com>
Date:   Wed Dec 28 15:29:11 2016 -0800

    Initial 8.x Commit
```
I think we might as well style it as part of this.

**Comment by jenbreese** [comment]
@pookmish Ok. I'll do that.

**Comment by jenbreese** [comment]
@kerri-augenstein  I'd like to style this 'unpublished page" text. How would you like it to look?
image

**Comment by sherakama** [comment]
haha yeah, that is a relic. 

The text doesn't have to stay. Not sure what my intentions with that were. Perhaps it could be swapped out with a yellow alert banner "This page is currently unpublished and not visible to the public."

**Comment by pookmish** [comment]
i like the alert banner idea. use what they are already some what familiar with to display the information.

**Comment by jenbreese** [comment]
@pookmish @sherakama I updated the text and added banner styling.

File: templates/node.html.twig [hunk 7]
```diff
@@ -34,7 +34,7 @@
       {% endif %}
 
       {% if not node.published %}
-        <p class="node--unpublished">{{ 'Unpublished'|t }}</p>
+        <p class="node--unpublished su-alert--warning">{{ 'This page is currently unpublished and not visible to the public.'|t }}</p>
```
Comment (sherakama) [inline templates/node.html.twig]: The alert template is available, why not use it?

```
{% 
set data = {
    'attributes': create_attribute({'class': 'su-alert su-alert--warning'}),
    'alert_text': "My message"|t,
  }
%}
{% include "@basic-dist/decanter/components/alert/alert.twig" with data %}
```

Comment (jenbreese) [reply templates/node.html.twig]: Indeed sir! :) all done.

File: src/scss/admin/_menu.scss [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Will need to make this less specific.
+nav {
+  li:hover {
+    .unpublished-indicator {
+      opacity: 1;
+    }
+  }
+}
```
Comment (sherakama) [inline src/scss/admin/_menu.scss]: I think this is fine. Not much you can do about that.

Comment (pookmish) [reply src/scss/admin/_menu.scss]: should it be `a:hover`. with this the indicator will show on all children of that link and all indicators of those children

Comment (sherakama) [reply src/scss/admin/_menu.scss]: That is a good point. I think you are probably right. 

Also, this doesn't appear to be working at all on hover at the moment.

Comment (jenbreese) [reply src/scss/admin/_menu.scss]: Yes, I was just about to write that.  I'll fix.

**Comment by jenbreese** [comment]
@sherakama and @pookmish I've fixed the hover.

File: src/scss/admin/_menu.scss [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  100% { opacity: 0; }
+}
+
+a:hover {
+  .unpublished-indicator {
+    display: inline-block;
+    opacity: 1;
+    animation: fade 5s ease-in-out;
+  }
+}
```
Comment (sherakama) [inline src/scss/admin/_menu.scss]:
```suggestion
a:hover,
a:focus {
  & > .unpublished-indicator {
    display: inline-block;
    opacity: 1;
    animation: fade 5s ease-in-out;
  }
}
```

Comment (sherakama) [reply src/scss/admin/_menu.scss]: Perhaps include the focus state and ensure it is only the direct unpublished indicator?

Comment (pookmish) [reply src/scss/admin/_menu.scss]: IMO the direct decedent isn't necessary and increases the amount of specificity that might not be wanted if we ever want to override the styles with a subtheme.

Comment (sherakama) [reply src/scss/admin/_menu.scss]: That is a fair point. With the activation on the `a` tag now it probably isn't needed and the less specificity is valued.

Comment (jenbreese) [reply src/scss/admin/_menu.scss]: Got it.

**Comment by sherakama** [comment]
Thank you.

**Pull request merged**